examples/ransomware_detection/stages/appshield_partitioner.py

```diff
@@ -0,0 +1,159 @@
+# Copyright (c) 2022-2023, NVIDIA CORPORATION.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import datetime
+import typing
+
+import mrc
+import pandas as pd
+from mrc.core import operators as ops
+
+from morpheus.cli.register_stage import register_stage
+from morpheus.config import Config
+from morpheus.config import PipelineModes
+from morpheus.messages.message_meta import AppShieldMessageMeta
+from morpheus.messages.message_meta import MessageMeta
+from morpheus.pipeline.multi_message_stage import MultiMessageStage
+from morpheus.pipeline.stream_pair import StreamPair
+
+
+@register_stage("appshield-partitioner", modes=[PipelineModes.FIL])
+class AppshieldPartitionerStage(MultiMessageStage):
+
+    def __init__(self,
+                 c: Config,
+                 interested_plugins: typing.List[str],
+                 cols_include: typing.List[str],
+                 plugins_schema: typing.Dict[str, any]):
+        super().__init__(c)
+
+        self._interested_plugins = interested_plugins
+        self._cols_include = cols_include
+        self._plugins_schema = plugins_schema
+        self._plugin_df_dict = {}
+
+    @property
+    def name(self) -> str:
+        return "appshield-partitioner"
+
+    def accepted_types(self) -> typing.Tuple:
+        """
+        Returns accepted input types for this stage.
+        """
+        return (MessageMeta, )
+
+    def supports_cpp_node(self):
+        return False
+
+    def _sources_with_multi_snapshots(self):
+        multiple_snapshots = {}
+        for source, source_dict in self._plugin_df_dict.items():
+            if len(source_dict) > 1:
+                for scan_id in source_dict.keys():
+                    multiple_snapshots.setdefault(source, []).append(scan_id)
+        return multiple_snapshots
+
+    def _hold_plugin_df(self, source, scan_id, plugin, plugin_df):
```

Comment:
Would it be possible to consider adding a time parameter to retain the plugin dataframes in memory? 

For instance, when the pipeline receives snapshot 1, plugin 1 processes it, and plugin 2 waits for plugin 3 to bundle its data and push it to the "create features" stage. However, in some cases, if plugin 3 fails to ingest data into the pipeline due to various reasons memory allocated to the plugin 1 and plugin 2 will not be released.

With the addition of a time parameter, we can ensure that plugin 1 and plugin 2 remain in memory only for the duration of the pipeline context, improving memory management and resource efficiency.

Reply:
Hi @bsuryadevara,
I think we already removing the old snapshots.
I printed the existing snapshots in the source (dictionary - memory object), and it looks like we already clean it whenever a new snapshot is added:
<img width="399" alt="dictionary_source" src="https://github.com/nv-morpheus/Morpheus/assets/33644618/88b05db5-1fb5-413f-87bb-8e952765e1f6">

Reply:
We have to handled the missed sequence snapshots? For example, if snapshots s1 and s12 are received but s3 is skipped and s4 arrives instead, will the workflow retain those snapshots held in the memory.
All we have to do is, If a sequence is found to be missing, add a check condition to initiate the cleanup process for the previous snapshots that are stored in memory.

Reply:
def _clean_snapshots(self, source, scan_id):
        scan_ids_exist = source.keys()
        for scan_id_exist in scan_ids_exist:
            if scan_id > scan_id_exist+2:
                del source[scan_id_exist]

    def _hold_plugin_df(self, source, scan_id, plugin, plugin_df):
        if source not in self._plugin_df_dict:
            self._plugin_df_dict[source] = {}
        source = self._plugin_df_dict[source]

        if scan_id not in source:
            source[scan_id] = {}
        snapshot = source[scan_id]

        if plugin not in snapshot:
            snapshot[plugin] = plugin_df
        else:
            snapshot[plugin] = pd.concat([snapshot[plugin], plugin_df])

        self._clean_snapshots(source, scan_id)